library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java

```diff
@@ -0,0 +1,289 @@
+	 * @return a new attribute
+	 */
+	static <R> RegistryEntryAttribute<R, Boolean> createBool(Registry<R> registry, Identifier id) {
+		return create(registry, id, Codec.BOOL, null);
+	}
+
+	/**
+	 * Creates a new {@code int} property.
+	 *
+	 * @param registry registry this attribute is tied to
+	 * @param id identifier
+	 * @param <R> registry entry type
+	 * @return a new attribute
+	 */
+	static <R> RegistryEntryAttribute<R, Integer> createInt(Registry<R> registry, Identifier id) {
+		return create(registry, id, Codec.INT, null);
+	}
+
+
+	/**
+	 * Creates a new {@code long} property.
+	 *
+	 * @param registry registry this attribute is tied to
+	 * @param id identifier
+	 * @param <R> registry entry type
+	 * @return a new attribute
+	 */
+	static <R> RegistryEntryAttribute<R, Long> createLong(Registry<R> registry, Identifier id) {
+		return create(registry, id, Codec.LONG, null);
+	}
+
+	/**
+	 * Creates a new {@code float} property.
+	 *
+	 * @param registry registry this attribute is tied to
+	 * @param id identifier
+	 * @param <R> registry entry type
+	 * @return a new attribute
+	 */
+	static <R> RegistryEntryAttribute<R, Float> createFloat(Registry<R> registry, Identifier id) {
+		return create(registry, id, Codec.FLOAT, null);
+	}
+
+	/**
+	 * Creates a new {@code double} property.
+	 *
+	 * @param registry registry this attribute is tied to
+	 * @param id identifier
+	 * @param <R> registry entry type
+	 * @return a new attribute
+	 */
+	static <R> RegistryEntryAttribute<R, Double> createDouble(Registry<R> registry, Identifier id) {
+		return create(registry, id, Codec.DOUBLE, null);
+	}
+
+	/**
+	 * Creates a new {@code String} property.
+	 *
+	 * @param registry registry this attribute is tied to
+	 * @param id identifier
+	 * @param <R> registry entry type
+	 * @return a new attribute
+	 */
+	static <R> RegistryEntryAttribute<R, String> createString(Registry<R> registry, Identifier id) {
+		return create(registry, id, Codec.STRING, null);
+	}
+
+	/**
+	 * Gets the registry this attribute is tied to.
+	 *
+	 * @return tied registry
+	 */
+	Registry<R> getRegistry();
+
+	/**
+	 * Gets the ID of this attribute.
+	 *
+	 * @return attribute ID
+	 */
+	Identifier getId();
+
+	/**
+	 * Gets the {@code Codec} used to (de)serialize this attribute's value.
+	 *
+	 * @return value codec
+	 */
+	Codec<V> getCodec();
+
+	/**
+	 * Gets the default value of this attribute. Can be {@code null}.
+	 *
+	 * @return default value
+	 */
+	@Nullable V getDefaultValue();
+
+	/**
+	 * Gets the value associated with this attribute for the specified entry.<p>
+	 *
+	 * If the item has no value for this attribute, the attribute's
+	 * {@linkplain #getDefaultValue() default value} will be returned instead, unless it is {@code null},
+	 * in which case an empty optional will be returned.
+	 *
+	 * @param entry registry entry
+	 * @return attribute value, or empty if no value is assigned
+	 */
+	default Optional<V> getValue(R entry) {
```

Comment:
I personally dislike Optionals, but my guess is there is a reason Optionals were chosen here?

Reply:
null bad

I don't remember honestly, I could just make this return `V`

Reply:
OH RIGHT, I made it `Optional` so primitive types are a bit easier to deal with (IE you don't need to interface with boxed types, just use an `orElseX` method)